base_layer/mmr/src/common.rs

```diff
@@ -75,7 +75,15 @@ pub fn find_peaks(size: usize) -> Vec<usize> {
             sum_prev_peaks += peak_size;
             num_left -= peak_size;
         }
-        peak_size >>= 1;
+        // need to verify if other peaks exist on the same height, in which case
+        // the size doesn't generate a complete mmr, e.g.
+        //    2
+        //   / \
+        //  0   1   3   4
+        // where we have two peaks at height 0.
```

Comment:
Ah! So the reason for the vec![] return is to say that the mmr size cannot exist.

Say we have an MMR of size 5 (nodes not leaf nodes), which would look like this
```
   2
 /   \
0    1  3    4
```

However that isnt a valid MMR size because it should look like this (7 nodes)

```text
        //        6
        //      /    \
        //    2      5
        //   / \    /  \
        //  0   1   3   4
```
As expected, `find_peaks(7) == [6]`

So perhaps we should return an Option here instead of an empty vec to force the caller to handle the case of an invalid size.

Reply:
I think the first example could make sense and might well represent a Merkle Mountain Range of size 5, and for that reason, we should be able to generate proofs for it. The implementations I have seen so far deal with any size MMR. But happy to make the `find_peaks` return an option in case, the size doesn't correspond to a fully generated MMR.

Reply:
A merkle mountain range cannot be size 5 because then it is malformed.

Reply:
The fact that it might break on `find_peaks` on invalid numbers, might point out that we might have an issue with the mmr, 
I agree here with @sdbondi an MMR cannot be for example size 5, 
And MMR containing 5 elements would be of size 8. 
```
     //              6
        //       /      \
        //     2        5
        //   / \      /  \
        //  0   1   3   4   7
```

It might be that our MMR implementation does not cap the peaks correctly

Reply:
I agree with the above comments. Let me refactor the code to output a `None` in case the MMR is not fully generated/complete/valid and check its usage later on. @SWvheerden in the example above, the size of the MMR is 8, spanning a 5 element set, I think.

Reply:
> Ah! So the reason for the vec![] return is to say that the mmr size cannot exist.
> 
> However that isnt a valid MMR size because it should look like this (7 nodes)
> 
> ```
>         //        6
>         //      /    \
>         //    2      5
>         //   / \    /  \
>         //  0   1   3   4
> ```
> 
> As expected, `find_peaks(7) == [6]`
> 
> So perhaps we should return an Option here instead of an empty vec to force the caller to handle the case of an invalid size. And add a comment explaining what `None` means.
> 

Yes, this. `find_peaks` should really be private and documented such that it should not be called before "bagging" the MMR. Returning an Option is ok too. I'll bet this wasn't done on day 1 on performance grounds and the assumption that it gets called after bagging (but should have been made private to make that assumption a contract). But @stringhandler and I discussed this and found that `Option` overheads aren't that high.